sanic/asgi.py

```diff
@@ -213,4 +213,5 @@ async def __call__(self) -> None:
             await self.sanic_app.handle_request(self.request)
         except Exception as e:
             await self.sanic_app.handle_exception(self.request, e)
-            self.stage = Stage.FAILED
+        finally:
+            self.stage = Stage.IDLE
```

Comment:
This might not be correct either. What if it goes to FAILED state during handler or exception? Then this sets it back to IDLE.

However, I cannot have a proper look at ASGI to see how it behaves, but I gather we should go FAILED only when a response is left without EOF. But no idea of how to actually handle that with ASGI servers. Needs some more investigation, in any case. Also prefer to keep this as close to http.py as possible.

Reply:
Would the lifecycle of the object (`Http` or `ASGI`, how do you guys call this type of object? Connection or something else?) be ended after `Stage.IDLE` or `stage.FAILED`? Can it accept further in-coming messages from the client? If it can, that means setting it to `Stage.IDLE` can make it accept more messages? 

By the way, I only see the stage being set to failed here in Sanic repo:
![image](https://user-images.githubusercontent.com/43905414/144733830-b9d51e5f-b745-4e90-bf33-23f3123b3d4f.png)

http.py, line 575
```python
    def respond(self, response: BaseHTTPResponse) -> BaseHTTPResponse:
        """
        Initiate new streaming response.

        Nothing is sent until the first send() call on the returned object, and
        calling this function multiple times will just alter the response to be
        given.
        """
        if self.stage is not Stage.HANDLER:
            self.stage = Stage.FAILED
            raise RuntimeError("Response already started")
```

Reply:
What about this?
```python
finally:
    if self.stage == Stage.RESPONSE:
        self.stage = Stage.IDLE
```

Reply:
I think maybe we don't need that in `__call__` at all, because it will be done in the send function, same in `http.py`

Reply:
I have a gut feeling that using try-finally is never good for this sort of operation. Can you set it IDLE where the final ASGI message (with EOF) is sent, and not in this function? Following the same flow as http.py is a good idea, and doing it in send() seems reasonable.

The meaning of `FAILED` is that nothing else can be done on the connection anymore, meaning that it needs to be disconnected. Not quite sure how that translates to ASGI, so cannot help you with that, but if possible, we should somehow tell the server that we are unable to make a valid response.

Reply:
> I have a gut feeling that using try-finally is never good for this sort of operation. Can you set it IDLE where the final ASGI message (with EOF) is sent, and not in this function? Following the same flow as http.py is a good idea, and doing it in send() seems reasonable.

Thanks, I think it has been done. ✅

> if possible, we should somehow tell the server that we are unable to make a valid response.

Yeah, I can definitely start looking to it. Since the current version worked somehow well on this, I think we can introduce this functionality in the future PR? 😂